### DISCUSSION TITLE
Is there an Offline Controller?

### DISCUSSION THREAD

**Discussion by troth530** [grblHAL/core#65] [Q&A]
Hi, does anyone use one or know of one in development.  SD card would be helpful.
Thanks!

**Comment by terjeio** [comment]
Something like [this](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack)?

Several drivers/boards has support for SD card streaming. It should also possible to write a plugin that adds a display and an encoder for selecting files to run for some of them.

**Comment by troth530** [comment]
Hi, @terjeio I've been looking into this.  It seems the MPG mode may work, but it seems to lack details on how to build it.  

I was thinking MKS or BTT TFT which are fully working senders for laser and CNC with SD built in.  Similar to what we have for an Ender 3 for example.  These can be connected through I2C (I believe) and can be interfaced through wifi apps.

grblHal opens up a huge amount of opportunity to improve CNC and laser work.  
We really appreciate your excellent work!!!!
